### PR TITLE
Jdbc analyzers state persistence

### PR DESCRIPTION
**Pull request by velramiir**
*Issue #35 *

*Description of changes:*
Add a JdbcStateProvider for the JdbcAnalyzers, works for memory and local disk

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/scala/com/amazon/deequ/analyzers/jdbc/IncrementalHistogramWithJdbc.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  withJdbc { connection =>
+
+    val table = Table("food_des", connection)
+
+    val analyzer = JdbcHistogram("fat_factor")
+
+    println(analyzer.calculate(table))
+
+    val stateProvider = JdbcFileSystemStateProvider("C:\\Users\\justu\\Documents\\histo")
```
Comment (phSch08) [inline src/main/scala/com/amazon/deequ/analyzers/jdbc/IncrementalHistogramWithJdbc.scala]: Please remove this

File: src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcStateProvider.scala [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+      case _: JdbcMinimum =>
+        persistDoubleState(state.asInstanceOf[MinState].minValue, identifier)
+
+      case _: JdbcMaximum =>
+        persistDoubleState(state.asInstanceOf[MaxState].maxValue, identifier)
+
+      case _: JdbcHistogram =>
+        persistFrequenciesLongState(state.asInstanceOf[JdbcFrequenciesAndNumRows], identifier)
+
```
Comment (phSch08) [inline src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcStateProvider.scala:106]: You may add Entropy here

File: src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcStateProvider.scala [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+      case _ : JdbcSum => SumState(loadDoubleState(identifier))
+
+      case _ : JdbcMean => loadDoubleLongState(identifier)
+
+      case _ : JdbcMinimum => MinState(loadDoubleState(identifier))
+
+      case _ : JdbcMaximum => MaxState(loadDoubleState(identifier))
+
+      case _ : JdbcHistogram => loadFrequenciesLongState(identifier)
+
```
Comment (phSch08) [inline src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcStateProvider.scala:143]: You may add Entropy here

**Review by phSch08 — CHANGES_REQUESTED** [review]

File: src/main/scala/com/amazon/deequ/analyzers/jdbc/IncrementalMaximumWithJdbc.scala [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    val table = Table("food_des", connection)
+
+    val analyzer = JdbcMaximum("fat_factor")
+
+    val maximum1 = analyzer.calculate(table)
+
+    val maxState = analyzer.computeStateFrom(table).getOrElse(MaxState(0))
+
+    val stateProvider = JdbcFileSystemStateProvider("")
```
Comment (phSch08) [inline src/main/scala/com/amazon/deequ/analyzers/jdbc/IncrementalMaximumWithJdbc.scala]:
```suggestion
    val stateProvider = JdbcFileSystemStateProvider("")// file system path to store at
```

**Review by phSch08 — APPROVED** [review]

**Review by phSch08 — APPROVED** [review]

File: src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcCorrelation.scala [hunk 5]
```diff
@@ -103,32 +104,29 @@ case class JdbcCorrelation(firstColumn: String,
 
     val result = statement.executeQuery()
 
-    try {
```
Comment (PaulZGit) [inline src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcCorrelation.scala]: Strange, why is try/catch missing here?

**Review by PaulZGit — COMMENTED** [review]
Strange: Already made changes are now newly in this branch/removed from master
Maybe we should talk to our tutors about that :O

File: src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcCorrelation.scala (hunk 5 above)
Comment (velramiir) [reply src/main/scala/com/amazon/deequ/analyzers/jdbc/JdbcCorrelation.scala]: because it's not included in any analyzer

**Pull request closed without merging**